forgejo: 1.18.2-1 -> 1.18.3-0

emilylange opened the pull request:
###### Description of changes

Release notes: https://codeberg.org/forgejo/forgejo/src/commit/109d8e57b164fabae7a42dc61ec58cb8a1f83200/RELEASE-NOTES.md#1-18-3-0

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

urandom2 commented on pkgs/applications/version-management/forgejo/default.nix:
```diff
@@ -59,12 +65,60 @@ buildGoModule rec {
+      coreutils
+      curl
+      jq
+      nix
+    ]}
+
+    releases=$(mktemp --suffix ".json")
+    curl "https://codeberg.org/api/v1/repos/forgejo/forgejo/releases?draft=false&pre-release=false&limit=1" \
+      --silent \
+      --header "accept: application/json" \
+      --output "$releases"
+
+    stable=$(mktemp --suffix ".json")
+    jq '.[0]
+      | .tag_name[1:] as $version
+      | ("forgejo-src-\($version).tar.gz") as $filename
+      | { $version, html_url } + (.assets | map(select(.name | startswith($filename)) | {(.name | split(".") | last): .browser_download_url}) | add)' \
+      "$releases" >"$stable"
+
+    version=$(jq -r .version "$stable")
+    release_url=$(jq -r .html_url "$stable")
+
+    if [[ "${version}" = "$version" ]]; then
+      echo "No new version found (already at $version)"
+      exit 0
+    fi
+
+    echo "Release: $release_url"
+
+    checksum_file=$(mktemp)
+    checksum_url=$(jq -r .sha256 "$stable")
+    curl "$checksum_url" --silent --output "$checksum_file"
+
+    sha256=$(cut --delimiter " " --fields 1 "$checksum_file")
+    sri_hash=$(nix hash to-sri --type sha256 "$sha256")
+
+    archive_url=$(jq -r .gz "$stable")
+
+    update-source-version "${pname}" "$version" "$sri_hash" "$archive_url"
+  '';
```
considering `writeShellApplication` does most of the setup for us, that seems like a better fit?:
```suggestion
  passthru.updateScript = let
    name = "update-forgejo";
  in  "${writeShellApplication {
    inherit name;
    runtimeInputs = [
      common-updater-scripts
      coreutils
      curl
      jq
      nix
    ];
    text = ''
      releases=$(mktemp --suffix ".json")
      curl "https://codeberg.org/api/v1/repos/forgejo/forgejo/releases?draft=false&pre-release=false&limit=1" \
        --silent \
        --header "accept: application/json" \
        --output "$releases"

      stable=$(mktemp --suffix ".json")
      jq '.[0]
        | .tag_name[1:] as $version
        | ("forgejo-src-\($version).tar.gz") as $filename
        | { $version, html_url } + (.assets | map(select(.name | startswith($filename)) | {(.name | split(".") | last): .browser_download_url}) | add)' \
        "$releases" >"$stable"

      version=$(jq -r .version "$stable")
      release_url=$(jq -r .html_url "$stable")

      if [[ "${version}" = "$version" ]]; then
        echo "No new version found (already at $version)"
        exit 0
      fi

      echo "Release: $release_url"

      checksum_file=$(mktemp)
      checksum_url=$(jq -r .sha256 "$stable")
      curl "$checksum_url" --silent --output "$checksum_file"

      sha256=$(cut --delimiter " " --fields 1 "$checksum_file")
      sri_hash=$(nix hash to-sri --type sha256 "$sha256")

      archive_url=$(jq -r .gz "$stable")

      update-source-version "${pname}" "$version" "$sri_hash" "$archive_url"
    '';
  }}/bin/${name}";
```

IndeedNotJames replied:
Good call!
I just exchanged the `let name = ""; in` with `lib.getExe` :)

urandom2 commented on pkgs/applications/version-management/forgejo/default.nix:
```diff
@@ -59,12 +65,60 @@ buildGoModule rec {
+
+    echo "Release: $release_url"
+
+    checksum_file=$(mktemp)
+    checksum_url=$(jq -r .sha256 "$stable")
+    curl "$checksum_url" --silent --output "$checksum_file"
+
+    sha256=$(cut --delimiter " " --fields 1 "$checksum_file")
+    sri_hash=$(nix hash to-sri --type sha256 "$sha256")
```
nit: this seems like it would read better with pipelines, and saves you from cleaning up the tempfile:
```suggestion
    sri_hash=$(nix hash to-sri --type sha256 "$(curl "$(jq -r .sha256 "$stable")" --silent | cut --delimiter " " --fields 1)")
```

IndeedNotJames replied:
I kept the `sha256` var, but replaced the tempfile with a *here string* :)

urandom2 commented on pkgs/applications/version-management/forgejo/default.nix:
```diff
@@ -59,12 +65,60 @@ buildGoModule rec {
+      --header "accept: application/json" \
+      --output "$releases"
+
+    stable=$(mktemp --suffix ".json")
+    jq '.[0]
+      | .tag_name[1:] as $version
+      | ("forgejo-src-\($version).tar.gz") as $filename
+      | { $version, html_url } + (.assets | map(select(.name | startswith($filename)) | {(.name | split(".") | last): .browser_download_url}) | add)' \
+      "$releases" >"$stable"
```
this should just drop in, but can we use a variable instead of a tempfile that needs to be cleaned up:
```suggestion
    stable="$(jq '.[0]
      | .tag_name[1:] as $version
      | ("forgejo-src-\($version).tar.gz") as $filename
      | { $version, html_url } + (.assets | map(select(.name | startswith($filename)) | {(.name | split(".") | last): .browser_download_url}) | add)' \
      "$releases")"
```

IndeedNotJames replied:
Replaced it with a *here string*, just like with `sha256` :)

urandom2 commented on pkgs/applications/version-management/forgejo/default.nix:
```diff
@@ -59,12 +65,60 @@ buildGoModule rec {
+      nix
+    ]}
+
+    releases=$(mktemp --suffix ".json")
+    curl "https://codeberg.org/api/v1/repos/forgejo/forgejo/releases?draft=false&pre-release=false&limit=1" \
+      --silent \
+      --header "accept: application/json" \
+      --output "$releases"
+
+    stable=$(mktemp --suffix ".json")
+    jq '.[0]
+      | .tag_name[1:] as $version
+      | ("forgejo-src-\($version).tar.gz") as $filename
+      | { $version, html_url } + (.assets | map(select(.name | startswith($filename)) | {(.name | split(".") | last): .browser_download_url}) | add)' \
+      "$releases" >"$stable"
```
nit: this seems like it would read better with pipelines, and saves you from cleaning up the tempfile:
```suggestion
    curl "https://codeberg.org/api/v1/repos/forgejo/forgejo/releases?draft=false&pre-release=false&limit=1" \
      --silent \
      --header "accept: application/json" \
      --output "$releases" \
    | jq '.[0]
      | .tag_name[1:] as $version
      | ("forgejo-src-\($version).tar.gz") as $filename
      | { $version, html_url } + (.assets | map(select(.name | startswith($filename)) | {(.name | split(".") | last): .browser_download_url}) | add)' \
      "$releases" >"$stable"
```

urandom2 reviewed (approved):
several nits and readability changes to the new `updateScript`, but the contents of the upgrade to the derivation are sane

emilylange committed:
forgejo: add `passthru.updateScript`

also add `meta.mainProgram`, since the resulting binary
is (currently?) called `gitea` not `forgejo`

Co-authored-by: Colin Arnott <urandom2@users.noreply.github.com>

emilylange committed:
forgejo: 1.18.2-1 -> 1.18.3-0

urandom2 reviewed (approved):
approval from before stands, `updateScript` looks nice, and thanks for the `lib.getExe` pointer!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/825
